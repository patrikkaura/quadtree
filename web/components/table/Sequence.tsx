import { useMemo } from "react";

import { groupNucleicPairs } from "./utils";

export type Props = {
  index: number;
  sequence: string;
};

const BASE_NUCLEIC_STYLE = {
  fontSize: 21,
  fontWeight: 500,
};

const BASE_GROUP_STYLE = {
  ...BASE_NUCLEIC_STYLE,
  borderBottom: "1px solid black",
};

export function Sequence({ index, sequence }: Props) {
  const grouped = groupNucleicPairs(sequence);

  const sequenceComponents = useMemo(
    () =>
      grouped.map((group, index) => {
        if (group.length > 1 && (group[0] === "C" || group[0] === "G")) {
          const color = group[0] === "C" ? "#6786C6" : "#ff5e7c";

          return (
            <span
              key={index}
              style={{
                ...BASE_GROUP_STYLE,
                color,
              }}
            >
              {group}
            </span>
          );
        } else if (
          group.length === 1 &&
          (group[0] === "G" || group[0] === "C")
        ) {
          const color = group[0] === "C" ? "#546ba8" : "#bf5a60";

          return (
            <span
              key={index}
              style={{
                ...BASE_NUCLEIC_STYLE,
                color,
              }}
            >
              {group}
            </span>
          );
        }

        return group;
      }),
    [grouped]
  );

  return <span data-testid={`sequence-${index}`}>{sequenceComponents}</span>;
}
