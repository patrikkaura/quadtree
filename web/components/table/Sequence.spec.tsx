import { render } from "@testing-library/react";

import { Props, Sequence } from "./Sequence";

describe("Sequence", () => {
  let props: Props = {
    index: 0,
    sequence: "ATATATAGGGGGGTATATATAGGGTATACCCCC",
  };

  it("should render", () => {
    const { container } = render(<Sequence {...props} />);
    expect(container).toMatchSnapshot();
  });
});
