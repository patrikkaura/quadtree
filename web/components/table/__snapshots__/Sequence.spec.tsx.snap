// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence should render 1`] = `
<div>
  <span
    data-testid="sequence-0"
  >
    A
    T
    A
    T
    A
    T
    A
    <span
      style="font-size: 22px; font-weight: 500; border-bottom: 1px solid black; color: rgb(255, 94, 124);"
    >
      G
      G
      G
      G
      G
      G
    </span>
    T
    A
    T
    A
    T
    A
    T
    A
    <span
      style="font-size: 22px; font-weight: 500; border-bottom: 1px solid black; color: rgb(255, 94, 124);"
    >
      G
      G
      G
    </span>
    T
    A
    T
    A
  </span>
</div>
`;
